consensus/polybft/fsm.go

```diff
@@ -0,0 +1,558 @@
+		index, exists := nodeIDIndexMap[commSeal.NodeID]
+		if !exists {
+			return fmt.Errorf("invalid node id = %s", commSeal.NodeID)
+		}
+
+		s, err := bls.UnmarshalSignature(commSeal.Signature)
+		if err != nil {
+			return fmt.Errorf("invalid signature = %s", commSeal.Signature)
+		}
+
+		signatures = append(signatures, s)
+
+		bitmap.Set(uint64(index))
+	}
+
+	aggregatedSignature, err := signatures.Aggregate().Marshal()
+	if err != nil {
+		return fmt.Errorf("could not aggregate seals: %w", err)
+	}
+
+	// include aggregated signature of all committed seals
+	// also includes bitmap which contains all indexes from validator set which provides there seals
+	extra.Committed = &Signature{
+		AggregatedSignature: aggregatedSignature,
+		Bitmap:              bitmap,
+	}
+
+	// Write extar data to header
+	f.block.Block.Header.ExtraData = append(make([]byte, 32), extra.MarshalRLPTo(nil)...)
+
+	if err := f.backend.CommitBlock(f.block); err != nil {
+		return err
+	}
+
+	if err := f.postInsertHook(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Height returns the height for the current round
+func (f *fsm) Height() uint64 {
+	return f.parent.Number + 1
+}
+
+// ValidatorSet returns the validator set for the current round
+func (f *fsm) ValidatorSet() pbft.ValidatorSet {
+	return f.validators
+}
+
+// IsStuck returns whether the pbft is stuck
+func (f *fsm) IsStuck(num uint64) (uint64, bool) {
+	return f.polybftBackend.CheckIfStuck(num)
+}
+
+// getValidatorSetDelta calculates validator set delta based on parent and current header
+func (f *fsm) getValidatorSetDelta(pendingBlockState *state.Transition) (*ValidatorSetDelta, error) {
+	provider := f.backend.GetStateProvider(pendingBlockState)
+	systemState := f.backend.GetSystemState(f.config, provider)
+	newValidators, err := systemState.GetValidatorSet()
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve validator set for current block %w", err)
+	}
+
+	return createValidatorSetDelta(f.logger, f.validators.Accounts(), newValidators), nil
+}
+
+// verifyValidatorsUptimeTx creates uptime transaction and compares its hash with the one extracted from the block.
+func (f *fsm) verifyValidatorsUptimeTx(transactions []*types.Transaction) error {
+	var blockUptimeTx *types.Transaction
+	if len(transactions) > 0 {
+		blockUptimeTx = transactions[0]
+	}
+
+	createdUptimeTx, err := f.createValidatorsUptimeTx()
+	if err != nil {
+		return err
+	}
+
+	if f.isEndOfEpoch {
+		if blockUptimeTx == nil {
+			return errors.New("uptime transaction is not found in the epoch ending block")
+		}
+
+		if blockUptimeTx.Hash != createdUptimeTx.Hash {
+			return fmt.Errorf(
+				"invalid uptime transaction. Expected '%s', but got '%s' uptime transaction hash",
+				blockUptimeTx.Hash,
+				createdUptimeTx.Hash,
+			)
+		}
+	} else {
+		if blockUptimeTx != nil && blockUptimeTx.Hash == createdUptimeTx.Hash {
+			return errors.New("didn't expect uptime transaction in the middle of an epoch")
+		}
+	}
```

Comment:
You can return from the previous condition, and just have this if below it, no need to drop it in the else statement